### PR TITLE
feat: redesign flake-ref parsing

### LINKED ISSUE #3: flakeref: Match Nix behavior for GitService `?ref=` and `?rev=`
We have a few differences at the moment:

- for `github:flox/runix/XXX`, XXX is always treated as a `rev`, when it could be a ref. Nix uses a regex match to determine which it is
- We have optional `ref` and optional `Pinned` (which contains optional `rev`), but Nix errors if a flakeref contains both `ref` and `rev`
- We serialize as `github:flox/runix?rev=XXX` but Nix serializes as `github:flox/runix/XXX`

### LINKED ISSUE #7: flakeref: support `?dir=`
- [ ] is this supported on `path:..` flakerefs as well?

### LINKED ISSUE #9: Implement Parsing for git flake refs
https://github.com/flox/runix/blob/3112401dc888111b7a074cbf4a52ebb2cd51e45b/crates/runix/src/flake_ref.rs#L137-L145

### LINKED ISSUE #16: Implement parsing for `tarball` and `file`
from the nix docs (https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-flake.html?highlight=flakeref#flake-references)

> In URL form, the schema must be `tarball+http://`, `tarball+https://` or `tarball+file://`. If the extension corresponds to a known archive format (.zip, `.tar`, `.tgz`, `.tar.gz`, `.tar.xz`, `.tar.bz2` or `.tar.zst`), then the `tarball+` can be dropped.

> In URL form, the schema must be `file+http://`, `file+https://` or `file+file://`. If the extension doesn’t correspond to a known archive format (as defined by the tarball fetcher), then the file+ prefix can be dropped.

The nix source code in https://github.com/NixOS/nix/blob/master/src/libfetchers/tarball.cc also reveals that tarball and file accept an `unpack` flag

### PR DESCRIPTION
**Pull request by ysndr**
This PR implements nix supported flakerefs.
Unlike the previous implementation all flakerefs are individual (generic) types.
This allows implementing converting to and from string representations on a per-type basis,
rather than necessarily covering all possible references in a single match statement.
As a side effect, implementing flake references one by one,
allowed us to match nix' behavior more closely (#3, #16).

Some flakerefs allow multiple protocols, but are essentially equivalent.
This includes `file`, `tarball`, `git` and "git forges" (e.g. `github`).
Such flakerefs have been implemented as generic types in order to share parsing logic,
and at the same time retain the ability to enforce individual origins statically.
For instance, it is now possible to define a composite type that requires local files (i.e. `git+file`, `path` `[file+]file://` or `[tarball+]file://`).

- [x] closes #7 
- [x] closes #3
- [x] closes #9
- [x] closes #16

### CODE REVIEW

**Comment by ysndr** [comment]
@mkenigs @notgne2 I invite you to add more test cases to make for coverage of flakerefs

File: crates/runix/src/flake_ref2/mod.rs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    Indirect(IndirectRef),
+    // /// https://cs.github.com/NixOS/nix/blob/f225f4307662fe9a57543d0c86c28aa9fddaf0d2/src/libfetchers/tarball.cc#L206
+    // Tarball(TarballRef),
+}
+
+impl FromStr for FlakeRef {
+    type Err = ParseFlakeRefError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let flake_ref = match s {
```
Comment (mkenigs) [inline crates/runix/src/flake_ref2/mod.rs]: @ysndr ❤️ 🦀

I would have gone for the good old else-if 😆

File: crates/runix/src/flake_ref2/mod.rs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+            _ if GitRef::<protocol::File>::parses(s) => s.parse::<GitRef<protocol::File>>()?.into(),
+            _ if GitRef::<protocol::SSH>::parses(s) => s.parse::<GitRef<protocol::SSH>>()?.into(),
+            _ if GitRef::<protocol::HTTP>::parses(s) => s.parse::<GitRef<protocol::HTTP>>()?.into(),
+            _ if GitRef::<protocol::HTTPS>::parses(s) => {
+                s.parse::<GitRef<protocol::HTTPS>>()?.into()
+            },
+            _ if IndirectRef::parses(s) => s.parse::<IndirectRef>()?.into(),
+            _ => Err(ParseFlakeRefError::Invalid)?,
+        };
+        Ok(flake_ref)
```
Comment (mkenigs) [inline crates/runix/src/flake_ref2/mod.rs]: Do you prefer this to wrapping the match with `Ok()`?

File: crates/runix/src/flake_ref2/file.rs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        let is_tarball_url = Path::new(url.path())
+            .file_name()
+            .map(|name| {
+                [
+                    ".zip", ".tar", ".tgz", ".tar.gz", ".tar.xz", ".tar.bz2", ".tar.zst",
+                ]
+                .iter()
+                .any(|ext| name.to_string_lossy().ends_with(ext))
+            })
+            .unwrap_or_default();
```
Comment (mkenigs) [inline crates/runix/src/flake_ref2/file.rs]: This could be a function also used in Tarball

File: crates/runix/src/flake_ref2/mod.rs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+#[derive(Serialize, Deserialize, Display, From, Debug)]
+#[serde(untagged)]
+pub enum FlakeRef {
+    FileFile(FileRef<protocol::File, file::File>),
+    FileHTTP(FileRef<protocol::HTTP, file::File>),
+    FileHTTPS(FileRef<protocol::HTTPS, file::File>),
+    TarballFile(FileRef<protocol::File, file::Tarball>),
+    TarballHTTP(FileRef<protocol::HTTP, file::Tarball>),
+    TarballHTTPS(FileRef<protocol::HTTPS, file::Tarball>),
```
Comment (mkenigs) [inline crates/runix/src/flake_ref2/mod.rs]: What's the advantage of making each of these a different type? I haven't gone through in enough detail yet to see if we're handling them all very differently, but I'm wondering why not `Tarball { protocol: enum Protocol { File, HTTP, HTTPS } }`

**Review by mkenigs — COMMENTED** [review]
Could you summarize the improvements this is making? Probably worth adding to the commit message. And why do we need it now?

File: crates/runix/src/flake_ref2/mod.rs (hunk 2 above)
Comment (ysndr) [reply crates/runix/src/flake_ref2/mod.rs]: yes, less indenting

File: crates/runix/src/flake_ref2/file.rs (hunk 3 above)
Comment (ysndr) [reply crates/runix/src/flake_ref2/file.rs]: Yeah i guess we can reuse that

**Comment by ysndr** [comment]
I'm still looking for nix' official test set, but all tests that existed with the old flake_ref work with the new ones and I added a couple of "roundtrips" to assure, we do not lose information on the way.

I stumbled over a very annoying [`serde` bug](https://github.com/serde-rs/serde/issues/1547#issuecomment-705744778) that basically says, `deny_unknown_fields` and `flatten` cannot be used together.

That and the fact that `serde_urlencode` is not self describing (which [triggers](https://github.com/nox/serde_urlencoded/issues/33) another [`serde` bug](https://github.com/serde-rs/serde/issues/1183) related to flattening), lets me wonder if we should use serde at all for query parsing or do the parsing manually, at the cost of legibility (https://github.com/flox/runix/pull/12/files#diff-fa82b2796286fd4622869172f2187af6691578ffbdf972e853826db2d4277fbcR200-R226).

File: crates/runix/src/flake_ref2/git_service.rs [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+            None => (rest, None),
+        };
+
+        let mut attributes: GitServiceAttributes =
+            serde_urlencoded::from_str(url.query().unwrap_or_default())?;
+
+        // if let Some(key) = attributes._unknown.keys().next() {
+        //     Err(ParseGitServiceError::UnkownAttribute(key.to_string()))?;
+        // }
+
```
Comment (ysndr) [inline crates/runix/src/flake_ref2/git_service.rs]:
```suggestion
```

File: crates/runix/src/flake_ref2/mod.rs (hunk 1 above)
Comment (ysndr) [reply crates/runix/src/flake_ref2/mod.rs]: image

File: crates/runix/Cargo.toml [hunk 6]
```diff
@@ -22,4 +22,9 @@ shell-escape = "0.1.5"
 tokio = { version = "1.21", features = ["full"] }
 tokio-stream = { version = "0.1.11", features = ["tokio-util", "io-util"] }
 thiserror = "1.0"
-url = "2.3"
+url = { version = "2.3",  features = [ "serde" ] }
+serde_urlencoded = "0.7.1"
+serde-value = "0.7.0"
```
Comment (ysndr) [inline crates/runix/Cargo.toml]:
```suggestion
```

File: crates/runix/src/flake_ref2/mod.rs (hunk 1 above)
Comment (mkenigs) [reply crates/runix/src/flake_ref2/mod.rs]: Lol

File: crates/runix/src/flake_ref2/mod.rs [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+    // /// https://cs.github.com/NixOS/nix/blob/f225f4307662fe9a57543d0c86c28aa9fddaf0d2/src/libfetchers/tarball.cc#L206
+    // Tarball(TarballRef),
+}
+
+impl FromStr for FlakeRef {
+    type Err = ParseFlakeRefError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let flake_ref = match s {
+            _ if FileRef::<protocol::File>::parses(s) => {
```
Comment (mkenigs) [inline crates/runix/src/flake_ref2/mod.rs]: This feels a little more bug prone than matching based on schema, because all the `parses()` functions aren't necessarily mutually exclusive, and if one of them has a bug it will potentially break every match arm after it. Maybe there's enough complexity here to make that necessary, but I do wonder if the previous approach was a bit more readable and less bug prone

Comment (ysndr) [reply crates/runix/src/flake_ref2/mod.rs]: thing is there is more complexity than just matching the prefix.

- see #16 for file and tarball
- we need to support `./path` and `/path` as well as `nixpkgs` -> `flake:nixpkgs`


sure if the first arm says "yup i take that" and then cant parse it, that will break the correct parser

i did have the issue with everything that failed parsing for any reason got parsed by `IndirectRef` before [ca47c9d](https://github.com/flox/runix/pull/15/commits/ca47c9d59354649682a27887eed2f275ef52d582)

So yeah the parses methods have to be strict enough to not intersect and also more flexible than just match statements

**Comment by ysndr** [comment]
I think this is good to go now

questions, that should not be blocking now:

- [ ] use a common `ParseError` type? Currently each reference implements its own parse error duplicating a couple of Error variants. We could instead just map to a single `ParseError` type that would have the union of all References' parse error variants.
- [ ] not use serde_urlencode to deserialize the query, but pick values manually from a map object. While, certainly more verbose, this might be able to work around the above mentioned serde limitations. which are very unfortunate

File: crates/runix/src/flake_ref2/mod.rs (hunk 4 above)
Comment (mkenigs) [reply crates/runix/src/flake_ref2/mod.rs]: @ysndr thoughts on this?

File: crates/runix/src/flake_ref/file.rs [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+    pub attributes: FileAttributes,
+}
+
+pub type FileRef<Protocol> = FileBasedRef<Protocol, application::File>;
+pub type TarballRef<Protocol> = FileBasedRef<Protocol, application::Tarball>;
+
+#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct FileAttributes {
+    #[serde(flatten)]
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/file.rs]: Having both `deny_unknown_fields` and `flatten` is misleading

File: crates/runix/src/flake_ref/file.rs [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+    #[derive(Default, Debug, Clone, PartialEq, Eq)]
+    pub struct File;
+    impl ApplicationProtocol for File {
+        fn protocol() -> Cow<'static, str> {
+            "file".into()
+        }
+
+        fn required(url: &Url) -> bool {
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/file.rs:143]: It's not very obvious to me what `required` does - I think it needs a comment or to be renamed

File: crates/runix/src/flake_ref/file.rs [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+        if maybe_ref.starts_with(&format!(
+            "{scheme_with_application}:",
+            scheme_with_application = Self::scheme()
+        )) {
+            return true;
+        }
+
+        if !Url::parse(maybe_ref)
+            .map(|url| Type::required(&url))
+            .unwrap_or_default()
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/file.rs]:
```suggestion
            .unwrap_or(false)
```

File: crates/runix/src/flake_ref/file.rs [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+
+    fn parses(maybe_ref: &str) -> bool {
+        if maybe_ref.starts_with(&format!(
+            "{scheme_with_application}:",
+            scheme_with_application = Self::scheme()
+        )) {
+            return true;
+        }
+
+        if !Url::parse(maybe_ref)
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/file.rs:196]: Can you add a comment? Or maybe this would be more clear if what `required` does was more clear?

File: crates/runix/src/flake_ref/file.rs [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+    }
+}
+
+impl<Protocol: FileProtocol, App: ApplicationProtocol> FromStr for FileBasedRef<Protocol, App> {
+    type Err = ParseFileError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let url = Url::parse(s)?;
+
+        // fix url
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/file.rs]: What does fixing a url mean?

File: crates/runix/src/flake_ref/file.rs [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+    type HttpsFileRef = super::FileRef<protocol::HTTPS>;
+
+    type FileTarballRef = super::TarballRef<protocol::File>;
+    type HttpTarballRef = super::TarballRef<protocol::HTTP>;
+    type HttpsTarballRef = super::TarballRef<protocol::HTTPS>;
+
+    #[test]
+    fn file_file_roundtrips() {
+        roundtrip::<FileFileRef>("file+file://somewhere/there");
+        roundtrip_to::<FileFileRef>("file://somewhere/there", "file+file://somewhere/there");
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/file.rs]: I wonder if file should be default and not serialized, so instead:
```
        roundtrip_to::<FileFileRef>("file+file://somewhere/there", "file://somewhere/there");
        roundtrip::<FileFileRef>("file://somewhere/there");
```

File: crates/runix/src/flake_ref/git_service.rs [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq)]
+#[serde_with::skip_serializing_none]
+#[serde(deny_unknown_fields)]
+pub struct GitServiceAttributes {
+    pub host: Option<String>,
+    pub dir: Option<PathBuf>,
+
+    #[serde(rename = "ref")]
+    pub reference: Option<String>,
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/git_service.rs:36]: Consistency is probably not that important, but if we care we could use `ref_` here to match `type_` https://github.com/flox/flox/blob/main/crates/flox-types/src/catalog/mod.rs#L40

File: crates/runix/src/flake_ref/git_service.rs [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+                Self::scheme().to_string(),
+                url.scheme().to_string(),
+            ));
+        }
+        let (owner, rest) = url
+            .path()
+            .split_once('/')
+            .ok_or(ParseGitServiceError::NoRepo)?;
+        let (repo, rev_or_ref) = match rest.split_once('/') {
+            Some((repo, rev)) => (repo, Some(rev.to_owned().into())),
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/git_service.rs]:
```suggestion
            Some((repo, rev_or_ref)) => (repo, Some(rev_or_ref.to_owned().into())),
```

File: crates/runix/src/flake_ref/indirect.rs [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+        if maybe_ref.starts_with("flake:") {
+            return true;
+        }
+
+        if maybe_ref.contains(':') {
+            return false;
+        }
+
+        ('a'..='z').any(|prefix| maybe_ref.starts_with(prefix))
+            || ('A'..='Z').any(|prefix| maybe_ref.starts_with(prefix))
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/indirect.rs:36]:
```suggestion
        ('a'..='z').any(|letter| maybe_ref.starts_with(letter))
            || ('A'..='Z').any(|letter| maybe_ref.starts_with(letter))
```

File: crates/runix/src/flake_ref/indirect.rs [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+        };
+
+        assert_eq!(IndirectRef::from_str("flake:nixpkgs").unwrap(), expected);
+        assert_eq!(IndirectRef::from_str("nixpkgs").unwrap(), expected);
+    }
+
+    #[test]
+    fn does_not_parse_other() {
+        IndirectRef::from_str("github:nixpkgs").unwrap_err();
+    }
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/indirect.rs:121]: Add a test with an attribute?

File: crates/runix/src/flake_ref/lock.rs [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+pub struct NarHash {
+    #[serde(rename = "narHash")]
+    pub nar_hash: String,
+}
+
+#[derive(Debug, Deserialize, Serialize, Deref, PartialEq, Eq, From, Clone)]
+#[from(forward)]
+pub struct LastModified {
+    #[serde(rename = "lastModified")]
+    pub last_modified: Timestamp,
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/lock.rs]: Is there a reason we need to use `LastModified` + flatten in structs instead of just using `Timestamp`?

File: crates/runix/src/flake_ref/path.rs [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+        };
+
+        if url.scheme() != Self::scheme() {
+            return Err(ParsePathRefError::InvalidScheme(
+                Self::scheme().to_string(),
+                url.scheme().to_string(),
+            ));
+        }
+        let path = Path::new(url.path()).to_path_buf();
+        let attributes: PathAttributes = dbg!(serde_urlencoded::from_str(dbg!(url
```
Comment (mkenigs) [inline crates/runix/src/flake_ref/path.rs]: Did you mean to leave `dbg!`? You've got a few in other places as well, not sure whether or not you want to keep them

**Review by mkenigs — CHANGES_REQUESTED** [review]

File: crates/runix/src/flake_ref/file.rs (hunk 8 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/file.rs]: Well technically not, but serde also has some issue with it as i've linked above.
I mean we DO want to deny fields that are not covered by the attributes, like `github:flox/runix?tef=feature/flakeref2` should fail with "unknown attribute 'tef'"...

File: crates/runix/src/flake_ref/file.rs (hunk 9 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/file.rs:143]: yes, will add a comment.

`required()` answers whether for the given URL, the application (file/tarball) has to be provided, or can be implied.
I.e. `required_for(url) = !implied_by(url)`

File: crates/runix/src/flake_ref/file.rs (hunk 10 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/file.rs]: don't like to guess what's the default for bool?
fine for me to make it explicit 😌

File: crates/runix/src/flake_ref/file.rs (hunk 12 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/file.rs]: resolve the application part from the url schema
filter out missing (required) or add (implied) application

File: crates/runix/src/flake_ref/file.rs (hunk 13 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/file.rs]: yeah we could check whether the application part is `required` before serializing
I'd like to serialize _more_ rather than losing info, but i'm happy doing whatever nix would do here..

File: crates/runix/src/flake_ref/git_service.rs (hunk 14 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/git_service.rs:36]: sure, either works I dont find it that much of an itch but we can rename this if you do

File: crates/runix/src/flake_ref/indirect.rs (hunk 16 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/indirect.rs:36]: I see where youre coming from with this but I cant help to read it as "latter" which kinda confuses me reading this..

File: crates/runix/src/flake_ref/indirect.rs (hunk 17 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/indirect.rs:121]: Yeah probably for the better..
would you mind otherwise i'll do on monday?

File: crates/runix/src/flake_ref/lock.rs (hunk 18 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/lock.rs]: Honestly it might be better to just use timestamp, yes
I made LastModified, so the timestamp and attribute name are typed but I agree thats just cognitve ballast..

File: crates/runix/src/flake_ref/path.rs (hunk 19 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/path.rs]: better not, they wont affect users but they will annoy us if they happen to be compiled if used as a library.

Comment (ysndr) [reply crates/runix/src/flake_ref/path.rs]: shall we remove them from the tests as well?

File: crates/runix/src/flake_ref/indirect.rs (hunk 16 above)
Comment (mkenigs) [reply crates/runix/src/flake_ref/indirect.rs:36]: No biggie, I just thought it was more clear since a prefix can be multiple characters and starts_with can accept a string

File: crates/runix/src/flake_ref/path.rs (hunk 19 above)
Comment (mkenigs) [reply crates/runix/src/flake_ref/path.rs]: No strong feelings about the tests 🤷

File: crates/runix/src/flake_ref/indirect.rs (hunk 17 above)
Comment (mkenigs) [reply crates/runix/src/flake_ref/indirect.rs:121]: Welp that's why we add tests, found a bug 😆 Pushed the fix as well. Looks like one other test is failing though

Comment (ysndr) [reply crates/runix/src/flake_ref/indirect.rs:121]: thanks, fixed

File: crates/runix/src/flake_ref/lock.rs (hunk 18 above)
Comment (ysndr) [reply crates/runix/src/flake_ref/lock.rs]: done

File: crates/runix/src/flake_ref2/mod.rs (hunk 4 above)
Comment (mkenigs) [reply crates/runix/src/flake_ref2/mod.rs]: Discussed and may be worth reworking at some point but not now

**Review by mkenigs — APPROVED** [review]
